fix(core): allow onDestroy unregistration while destroying

garrettld opened the pull request:
Ensure that all onDestroy callbacks are called by allowing unregistering of onDestroy hooks while destroying.

Fixes #50221

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
`onDestroy` callbacks that unregister themselves when called will cause the next `onDestroy` callback to be skipped.

Issue Number: #50221


## What is the new behavior?
All `onDestroy` callbacks that are present when the destroy process starts will be called, regardless of any unregistration that may occur while destroying.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

garrettld committed:
fix(core): allow onDestroy unregistration while destroying

Ensure that all onDestroy callbacks are called by allowing unregistering of onDestroy hooks while destroying.

Fixes #50221

garrettld committed:
fixup! fix(core): allow onDestroy unregistration while destroying

garrettld commented:
@pkozlowski-opensource I went with setting the ON_DESTROY_HOOKS value to null (per your suggestion) to avoid the relatively more expensive approach of copying the array. Thanks for letting me take a crack at this.

AndrewKushnir commented on packages/core/src/render3/node_manipulation.ts:
```diff
@@ -466,6 +466,7 @@ function processCleanups(tView: TView, lView: LView): void {
     lView[CLEANUP] = null;
   }
   const destroyHooks = lView[ON_DESTROY_HOOKS];
+  lView[ON_DESTROY_HOOKS] = null;
```
It looks like there is a similar call below (at line 475), should we remove it from there? I think it'd also be helpful to add a comment into the code on why we reset it before invoking callbacks.

AndrewKushnir reviewed (commented):
@garrettld thanks for the PR. I've just left a quick comment.

I'd also like to ask @pkozlowski-opensource to take a look, he is more familiar with this part of the code.

garrettld committed:
fixup! fix(core): allow onDestroy unregistration while destroying

garrettld commented:
@AndrewKushnir I removed the second `lView[ON_DESTROY_HOOKS] = null;`, I think it's relatively safe to do so since user code should not be able to register new hooks at this point since `LViewFlags.Destroyed` is already set. I also moved the reset into the `if (destroyHooks !== null)` block since it's only necessary if that condition is true.

pkozlowski-opensource commented:
> Thanks for letting me take a crack at this.

Thank YOU for contributing the fix, appreciate it!

AndrewKushnir commented:
[Presubmit](http://test/OCL:531051235:BASE:531051261:1683765552595:8adbd681).

pkozlowski-opensource commented:
@garrettld I've just realised that there is another code path that suffers from the same problem, this test (a bit convoluted, bit still pretty simple) exposes it:

```typescript
it('should allow unregistration while destroying', () => {
      const destroyedLog: string[] = [];

      const envInjector = createEnvironmentInjector([], TestBed.inject(EnvironmentInjector));
      const destroyRef = envInjector.get(DestroyRef);

      const unregister = destroyRef.onDestroy(() => {
        destroyedLog.push('first');
        unregister();
      });
      destroyRef.onDestroy(() => {
        destroyedLog.push('second');
      });

      expect(destroyedLog).toEqual([]);

      envInjector.destroy();
      expect(destroyedLog).toEqual(['first', 'second']);
    });
```

A similar fix needs to be done in https://github.com/angular/angular/blob/8ce99f5a0d958d933e4dda8036abc3683b61a9a7/packages/core/src/di/r3_injector.ts#L202 

Do you think you could update the PR to include the above test + fix for the other code path?

garrettld committed:
fixup! fix(core): allow onDestroy unregistration while destroying

pkozlowski-opensource commented:
Thnx for the update @garrettld - looks great!

pkozlowski-opensource commented:
caretaker note: the latest [presubmit](https://fusion2.corp.google.com/presubmit/531187338/OCL:531187338:BASE:531187365:1683812853692:903635fb/targets) failures are pre-existing.

AndrewKushnir commented:
This PR was merged into the repository by commit 03f1e240b368c4c87fd8412605bf0eff9adc153d.
